// @kie/mock-github/build/src/github/schema/history.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MergeSchema = exports.PushSchema = void 0;
const repository_history_types_1 = require("../../github/repository/history/repository-history.types");
const repository_file_1 = require("../../github/schema/repository-file");
exports.PushSchema = {
    type: "object",
    properties: {
        branch: {
            type: "string",
        },
        files: {
            ...repository_file_1.CreateRepositoryFileSchema,
            nullable: true,
        },
        commitMessage: {
            type: "string",
            nullable: true,
        },
        action: {
            type: "string",
            enum: [repository_history_types_1.GitActionTypes.MERGE, repository_history_types_1.GitActionTypes.PUSH],
        },
    },
    required: ["branch", "action"],
};
exports.MergeSchema = {
    type: "object",
    properties: {
        head: {
            type: "string",
        },
        base: {
            type: "string",
        },
        commitMessage: {
            type: "string",
            nullable: true,
        },
        action: {
            type: "string",
            enum: [repository_history_types_1.GitActionTypes.MERGE, repository_history_types_1.GitActionTypes.PUSH],
        },
    },
    required: ["head", "base", "action"],
};
