// @kie/mock-github/build/src/github/schema/repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RepositorySchema = void 0;
const branch_1 = require("../../github/schema/branch");
const history_1 = require("../../github/schema/history");
const repository_file_1 = require("../../github/schema/repository-file");
exports.RepositorySchema = {
    type: "object",
    properties: {
        localBranches: {
            ...branch_1.BranchSchema,
            nullable: true,
        },
        pushedBranches: {
            ...branch_1.BranchSchema,
            nullable: true,
        },
        currentBranch: {
            type: "string",
            nullable: true,
        },
        owner: {
            type: "string",
            nullable: true,
        },
        forkedFrom: {
            type: "string",
            nullable: true,
        },
        files: {
            ...repository_file_1.CreateRepositoryFileSchema,
            nullable: true,
        },
        history: {
            type: "array",
            items: {
                anyOf: [history_1.MergeSchema, history_1.PushSchema],
            },
            nullable: true,
        },
    },
};
