// @kie/act-js/build/src/step-mocker/step-mocker.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isStepIdentifierUsingBeforeOrAfter = exports.isStepIdentifierUsingAfter = exports.isStepIdentifierUsingBefore = exports.isStepIdentifierUsingIndex = exports.isStepIdentifierUsingRun = exports.isStepIdentifierUsingUses = exports.isStepIdentifierUsingId = exports.isStepIdentifierUsingName = void 0;
function isStepIdentifierUsingName(step) {
    return Object.prototype.hasOwnProperty.call(step, "name");
}
exports.isStepIdentifierUsingName = isStepIdentifierUsingName;
function isStepIdentifierUsingId(step) {
    return Object.prototype.hasOwnProperty.call(step, "id");
}
exports.isStepIdentifierUsingId = isStepIdentifierUsingId;
function isStepIdentifierUsingUses(step) {
    return Object.prototype.hasOwnProperty.call(step, "uses");
}
exports.isStepIdentifierUsingUses = isStepIdentifierUsingUses;
function isStepIdentifierUsingRun(step) {
    return Object.prototype.hasOwnProperty.call(step, "run");
}
exports.isStepIdentifierUsingRun = isStepIdentifierUsingRun;
function isStepIdentifierUsingIndex(step) {
    return Object.prototype.hasOwnProperty.call(step, "index");
}
exports.isStepIdentifierUsingIndex = isStepIdentifierUsingIndex;
function isStepIdentifierUsingBefore(step) {
    return Object.prototype.hasOwnProperty.call(step, "before");
}
exports.isStepIdentifierUsingBefore = isStepIdentifierUsingBefore;
function isStepIdentifierUsingAfter(step) {
    return Object.prototype.hasOwnProperty.call(step, "after");
}
exports.isStepIdentifierUsingAfter = isStepIdentifierUsingAfter;
function isStepIdentifierUsingBeforeOrAfter(step) {
    return isStepIdentifierUsingBefore(step) || isStepIdentifierUsingAfter(step);
}
exports.isStepIdentifierUsingBeforeOrAfter = isStepIdentifierUsingBeforeOrAfter;
